src/libtego/source/ed25519.cpp

```diff
@@ -71,36 +71,60 @@ tego_v3_onion_service_id::tego_v3_onion_service_id(
     size_t serviceIdStringLength)
 {
     TEGO_THROW_IF_NULL(serviceIdString);
-    TEGO_THROW_IF_FALSE(serviceIdStringLength >= TEGO_V3_ONION_SERVICE_ID_LENGTH);
+    TEGO_THROW_IF_FALSE(is_valid(serviceIdString, serviceIdStringLength));
+
+    // copy to our internal buffer
+    std::copy(
+        &serviceIdString[0],
+        &serviceIdString[TEGO_V3_ONION_SERVICE_ID_LENGTH],
+        this->data);
+}
+
+tego_bool_t tego_v3_onion_service_id::is_valid(
+    const char* serviceIdString,
+    size_t serviceIdStringLength)
+{
+    logger::trace();
+    TEGO_THROW_IF_NULL(serviceIdString);
+    logger::trace();
+    if (serviceIdStringLength < TEGO_V3_ONION_SERVICE_ID_LENGTH) { /* should this throw? */
+        return TEGO_FALSE;
+    }
+    logger::trace();
 
     std::string_view serviceIdView(serviceIdString, TEGO_V3_ONION_SERVICE_ID_LENGTH);
-    uint8_t rawServiceId[TEGO_V3_ONION_SERVICE_ID_RAW_SIZE] = {0};
+    uint8_t decodedServiceId[TEGO_V3_ONION_SERVICE_ID_RAW_SIZE] = {0};
 
     // base32 decode service serviceId
     const auto bytesDecoded = ::base32_decode(
-        reinterpret_cast<char*>(rawServiceId),
-        sizeof(rawServiceId),
+        reinterpret_cast<char *>(decodedServiceId),
+        sizeof(decodedServiceId),
         serviceIdView.data(),
         serviceIdView.size());
-    TEGO_THROW_IF_FALSE(bytesDecoded == sizeof(rawServiceId));
 
-    // verify correct version byte
-    TEGO_THROW_IF_FALSE(rawServiceId[TEGO_V3_ONION_SERVICE_ID_VERSION_OFFSET] == 0x03);
+    logger::trace();
+    // check successful base32 decode and correct version byte
+    if (bytesDecoded != sizeof(decodedServiceId) ||
+        decodedServiceId[TEGO_V3_ONION_SERVICE_ID_VERSION_OFFSET] != 0x03) {
+        return TEGO_FALSE;
+    }
+    logger::trace();
 
-    // first part of the rawServiceId is the publicKey
-    auto& rawPublicKey = reinterpret_cast<uint8_t (&)[ED25519_PUBKEY_LEN]>(rawServiceId);
+    auto& rawPublicKey = reinterpret_cast<uint8_t (&)[ED25519_PUBKEY_LEN]>(decodedServiceId);
 
     // calculate the truncated checksum for the public key
     uint8_t truncatedChecksum[TEGO_V3_ONION_SERVICE_ID_CHECKSUM_SIZE] = {0};
     tego::truncated_checksum_from_ed25519_public_key(truncatedChecksum, rawPublicKey);
+    logger::trace();
 
     // verify the first two bytes of checksum in service id match our calculated checksum
-    TEGO_THROW_IF_FALSE(
-        rawServiceId[TEGO_V3_ONION_SERVICE_ID_CHECKSUM_OFFSET    ] == truncatedChecksum[0] &&
-        rawServiceId[TEGO_V3_ONION_SERVICE_ID_CHECKSUM_OFFSET + 1] == truncatedChecksum[1]);
+    if( decodedServiceId[TEGO_V3_ONION_SERVICE_ID_CHECKSUM_OFFSET    ] != truncatedChecksum[0] ||
```

Comment:
this is correct but harder to reason about (for me at least). Please save the result of the check into a temp bool, and verify that val is true

Reply:
is the new version better?